File: clients/python/tests/e2e/_tck_provenance.py
"""Provenance tagging for e2e tests.

Every conversation, entity, and reasoning step the e2e suite creates gets
tagged with metadata that traces it back to:

    - the language client that ran the test  (tck_client)
    - the specific test that created it      (tck_test)
    - the GitHub Actions run that triggered it (tck_run_id, tck_run_attempt)
    - the commit SHA + branch                (tck_sha, tck_branch)
    - the suite start time                   (tck_started_at)
    - the hostname / runner                  (tck_host)

Querying provenance after the fact (with workspace-admin scope, via Cypher):

    # All e2e data from a single CI run, regardless of language:
    MATCH (c:Conversation) WHERE c.metadata.tck_run_id = '12345' RETURN c

    # Every entity tagged by the python suite:
    MATCH (e:Entity) WHERE e.description STARTS WITH '[tck:python' RETURN e

    # All reasoning steps the test harness emitted:
    MATCH (s:AgentStep) WHERE s.reasoning STARTS WITH 'TCK e2e' RETURN s
"""

from __future__ import annotations

import json
import os
import socket
from datetime import datetime, timezone
from functools import lru_cache
from typing import Any


CLIENT_NAME = "python"


@lru_cache(maxsize=1)
def run_info() -> dict[str, str]:
    """Process-level metadata that's identical for every test in this run."""
    sha = (os.environ.get("GITHUB_SHA") or "local")[:7]
    return {
        "tck_client": CLIENT_NAME,
        "tck_run_id": os.environ.get("GITHUB_RUN_ID") or "local",
        "tck_run_attempt": os.environ.get("GITHUB_RUN_ATTEMPT") or "1",
        "tck_workflow": os.environ.get("GITHUB_WORKFLOW") or "local",
        "tck_sha": sha,
        "tck_branch": os.environ.get("GITHUB_REF_NAME") or "local",
        "tck_started_at": datetime.now(timezone.utc).isoformat(),
        "tck_host": os.environ.get("RUNNER_NAME") or socket.gethostname(),
    }


def metadata_for(test_name: str, **extra: Any) -> dict[str, Any]:
    """Per-test metadata: run_info plus the test name and any extras.

    Renamed from `test_metadata` to avoid pytest collecting helpers prefixed
    with `test_` as test cases.
    """
    return {**run_info(), "tck_test": test_name, **extra}


def tag_description(test_name: str, description: str) -> str:
    """Prefix an entity description with a searchable provenance tag."""
    info = run_info()
    return f"[tck:{info['tck_client']}:{info['tck_run_id']}:{test_name}] {description}"


def provenance_reasoning(test_name: str, phase: str = "setup") -> str:
    """Reasoning text used when recording a provenance step on a conversation."""
    info = run_info()
    return (
        f"TCK e2e test {phase}: {test_name} "
        f"[client={info['tck_client']}, run={info['tck_run_id']}, "
        f"sha={info['tck_sha']}, branch={info['tck_branch']}]"
    )


def provenance_result(test_name: str, **extra: Any) -> str:
    """JSON-encoded metadata blob suitable for the `result` field of a step."""
    return json.dumps(metadata_for(test_name, **extra), separators=(",", ":"))

File: clients/csharp/tests/Neo4j.AgentMemory.Tests/E2E/TckProvenance.cs
using System.Text.Json;
using Neo4j.AgentMemory.Models;

namespace Neo4j.AgentMemory.Tests.E2E;

/// <summary>
/// Provenance tagging for e2e tests.
///
/// Every conversation, entity, and reasoning step the e2e suite creates is
/// tagged with metadata that traces it back to:
///
///   • the language client            (tck_client)
///   • the specific test              (tck_test)
///   • the GitHub Actions run         (tck_run_id, tck_run_attempt)
///   • the commit SHA + branch        (tck_sha, tck_branch)
///   • the suite start time           (tck_started_at)
///   • the runner / hostname          (tck_host)
///
/// Querying provenance after the fact (with workspace-admin Cypher access):
///
///   MATCH (c:Conversation) WHERE c.metadata.tck_run_id = '12345' RETURN c
///   MATCH (e:Entity) WHERE e.description STARTS WITH '[tck:csharp' RETURN e
///   MATCH (s:AgentStep) WHERE s.reasoning STARTS WITH 'TCK e2e' RETURN s
/// </summary>
public static class TckProvenance
{
    public const string ClientName = "csharp";

    private static readonly Lazy<Dictionary<string, object?>> _runInfo =
        new(() =>
        {
            var sha = Environment.GetEnvironmentVariable("GITHUB_SHA") ?? "local";
            if (sha.Length > 7) sha = sha[..7];
            var host = Environment.GetEnvironmentVariable("RUNNER_NAME")
                ?? Environment.MachineName;
            return new Dictionary<string, object?>
            {
                ["tck_client"] = ClientName,
                ["tck_run_id"] = Environment.GetEnvironmentVariable("GITHUB_RUN_ID") ?? "local",
                ["tck_run_attempt"] = Environment.GetEnvironmentVariable("GITHUB_RUN_ATTEMPT") ?? "1",
                ["tck_workflow"] = Environment.GetEnvironmentVariable("GITHUB_WORKFLOW") ?? "local",
                ["tck_sha"] = sha,
                ["tck_branch"] = Environment.GetEnvironmentVariable("GITHUB_REF_NAME") ?? "local",
                ["tck_started_at"] = DateTime.UtcNow.ToString("o"),
                ["tck_host"] = host,
            };
        });

    public static IReadOnlyDictionary<string, object?> RunInfo => _runInfo.Value;

    public static Dictionary<string, object?> MetadataFor(string testName, IDictionary<string, object?>? extra = null)
    {
        var output = new Dictionary<string, object?>(_runInfo.Value);
        output["tck_test"] = testName;
        if (extra != null)
        {
            foreach (var kv in extra) output[kv.Key] = kv.Value;
        }
        return output;
    }

    public static string TagDescription(string testName, string description)
    {
        var info = _runInfo.Value;
        return $"[tck:{info["tck_client"]}:{info["tck_run_id"]}:{testName}] {description}";
    }

    public static string ReasoningText(string testName, string phase = "setup")
    {
        var info = _runInfo.Value;
        return $"TCK e2e test {phase}: {testName} " +
               $"[client={info["tck_client"]}, run={info["tck_run_id"]}, " +
               $"sha={info["tck_sha"]}, branch={info["tck_branch"]}]";
    }

    public static string ResultJson(string testName, IDictionary<string, object?>? extra = null)
    {
        return JsonSerializer.Serialize(MetadataFor(testName, extra));
    }

    /// <summary>
    /// Best-effort: record a reasoning step on the conversation tying it
    /// back to the originating test. Never throws.
    /// </summary>
    public static async Task RecordStepAsync(
        MemoryClient client, string conversationId, string testName,
        string phase = "setup", string action = "create_conversation")
    {
        try
        {
            await client.Reasoning.RecordStepAsync(
                conversationId,
                reasoning: ReasoningText(testName, phase),
                actionTaken: action,
                result: ResultJson(testName, new Dictionary<string, object?>
                {
                    ["conversation_id"] = conversationId,
                }));
        }
        catch
        {
            // Provenance is best-effort.
        }
    }
}
